contracts/utils.go

```diff
@@ -4,85 +4,147 @@ import (
 	"bytes"
 	"errors"
 	"math/big"
+	"time"
 
-	abipkg "github.com/celo-org/celo-blockchain/accounts/abi"
+	"github.com/celo-org/celo-blockchain/accounts/abi"
 	"github.com/celo-org/celo-blockchain/common"
 	"github.com/celo-org/celo-blockchain/common/hexutil"
 	"github.com/celo-org/celo-blockchain/core/vm"
 	"github.com/celo-org/celo-blockchain/log"
+	"github.com/celo-org/celo-blockchain/metrics"
 )
 
-// systemCaller is the caller when the EVM is invoked from the within the blockchain system.
-var systemCaller = vm.AccountRef(common.HexToAddress("0x0"))
+// SystemCaller is the caller when the EVM is invoked from the within the blockchain system.
+var SystemCaller = common.HexToAddress("0x0")
 
-func StaticCallFromSystem(evm *vm.EVM, contractAddress common.Address, abi abipkg.ABI, funcName string, args []interface{}, returnObj interface{}, gas uint64) (uint64, error) {
-	staticCall := func(transactionData []byte) ([]byte, uint64, error) {
-		return evm.StaticCall(systemCaller, contractAddress, transactionData, gas)
-	}
+type Contract struct {
+	abi     *abi.ABI
+	address common.Address
+	caller  common.Address
+}
 
-	return handleABICall(evm, abi, funcName, args, returnObj, staticCall)
+type Backend interface {
+	Call(caller vm.ContractRef, addr common.Address, input []byte, gas uint64, value *big.Int) (ret []byte, leftOverGas uint64, err error)
+	StaticCall(caller vm.ContractRef, addr common.Address, input []byte, gas uint64) (ret []byte, leftOverGas uint64, err error)
 }
 
-func CallFromSystem(evm *vm.EVM, contractAddress common.Address, abi abipkg.ABI, funcName string, args []interface{}, returnObj interface{}, gas uint64, value *big.Int) (uint64, error) {
-	call := func(transactionData []byte) ([]byte, uint64, error) {
-		return evm.Call(systemCaller, contractAddress, transactionData, gas, value)
+func NewContract(contractAbi *abi.ABI, contractAddress, defaultCaller common.Address) *Contract {
+	return &Contract{
+		abi:     contractAbi,
+		address: contractAddress,
+		caller:  defaultCaller,
 	}
-	return handleABICall(evm, abi, funcName, args, returnObj, call)
 }
 
-var (
-	errorSig     = []byte{0x08, 0xc3, 0x79, 0xa0} // Keccak256("Error(string)")[:4]
-	abiString, _ = abipkg.NewType("string", "", nil)
-)
+type ExecOpts struct {
+	Backend Backend
+	MaxGas  uint64
+	Value   *big.Int
```

Comment:
I think either `Value` doesn't belong here, or `method` and `args` should also be included here, because they're pieces of the same thing (what the call is which we are doing).  To a lesser degree I think the same goes for `MaxGas` (either it doesn't belong here or `method` and `args` should be here too)... to a lesser degree because it could be considered more of a constraint on how the call is to be done rather than what the call is which should be done.